crates/autopilot/src/domain/competition/participation_guard/db.rs

```diff
@@ -0,0 +1,123 @@
+use {
+    crate::{
+        database::Postgres,
+        domain::{eth, Metrics},
+        infra,
+    },
+    ethrpc::block_stream::CurrentBlockWatcher,
+    std::{
+        collections::HashMap,
+        sync::Arc,
+        time::{Duration, Instant},
+    },
+};
+
+/// Checks the DB by searching for solvers that won N last consecutive auctions
+/// but never settled any of them.
+#[derive(Clone)]
+pub(super) struct Validator(Arc<Inner>);
+
+struct Inner {
+    db: Postgres,
+    banned_solvers: dashmap::DashMap<eth::Address, Instant>,
+    ttl: Duration,
+    last_auctions_count: u32,
+    drivers_by_address: HashMap<eth::Address, Arc<infra::Driver>>,
+}
+
+impl Validator {
+    pub fn new(
+        db: Postgres,
+        current_block: CurrentBlockWatcher,
+        settlement_updates_receiver: tokio::sync::mpsc::UnboundedReceiver<()>,
+        ttl: Duration,
+        last_auctions_count: u32,
+        drivers_by_address: HashMap<eth::Address, Arc<infra::Driver>>,
+    ) -> Self {
+        let self_ = Self(Arc::new(Inner {
+            db,
+            banned_solvers: Default::default(),
+            ttl,
+            last_auctions_count,
+            drivers_by_address,
+        }));
+
+        self_.start_maintenance(settlement_updates_receiver, current_block);
+
+        self_
+    }
+
+    /// Update the internal cache only once the settlement table is updated to
+    /// avoid redundant DB queries.
+    fn start_maintenance(
+        &self,
+        mut settlement_updates_receiver: tokio::sync::mpsc::UnboundedReceiver<()>,
+        current_block: CurrentBlockWatcher,
+    ) {
+        let self_ = self.clone();
+        tokio::spawn(async move {
+            while settlement_updates_receiver.recv().await.is_some() {
+                let current_block = current_block.borrow().number;
+                match self_
+                    .0
+                    .db
+                    .find_non_settling_solvers(self_.0.last_auctions_count, current_block)
+                    .await
+                {
+                    Ok(non_settling_solvers) => {
+                        let non_settling_drivers = non_settling_solvers
+                            .into_iter()
+                            .filter_map(|solver| {
+                                let address = eth::Address(solver.0.into());
```

Comment:
If you move `find_non_settling_solvers` to `infra::Persistence` this line won't be needed.